onnxruntime/core/providers/cpu/optional/optional_ops.cc

```diff
@@ -0,0 +1,178 @@
+
+    } else {
+      // Will not reach here
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Only Optional type OrtValues containing Tensors "
+                             "and Sequence Tensors can be used as input to "
+                             "OptionalGetElement op");
+    }
+  } else {  // No input was provided - we use the type proto to construct the output OrtValue
+
+    CheckValidTypeProto(*type_proto_);
+
+    // type is either Tensor or TensorSeq (we have validated this already in CheckValidTypeProto())
+    if (type_proto_->has_tensor_type()) {
+      ctx->OutputOptionalWithoutData<Tensor>(0);
+    } else {
+      ctx->OutputOptionalWithoutData<TensorSeq>(0);
+    }
+  }
+
+  return Status::OK();
+}
+
+Status OptionalHasElement::Compute(OpKernelContext* ctx) const {
+  const auto* input_ort_value = ctx->GetInputOrtValue(0);
+
+  // Output is a scalar
+  auto* output_tensor = ctx->Output(0, {});
+  output_tensor->MutableData<bool>()[0] = input_ort_value->IsAllocated();
+
+  return Status::OK();
+}
+
+Status OptionalGetElement::Compute(OpKernelContext* ctx) const {
+  const auto* input_ort_value = ctx->GetInputOrtValue(0);
+
+  if (!input_ort_value->IsAllocated()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Trying to use OptionalGetElement on an optional type "
+                           "OrtValue which contains no data");
+  }
+
+  if (input_ort_value->IsTensor()) {
+    const auto* input_tensor = &input_ort_value->Get<Tensor>();
+    auto* output_tensor = ctx->Output(0, input_tensor->Shape());
+
+    // If the allocation planner had deemed that we re-use the input OrtValue
+    // as the output OrtValue, the data pointers in the input_tensor and the
+    // output_tensor will be the same and the copy is a no-op.
+    // CopyCpuTensor() already has such copy optimizations - so
+    // just re-ue it.
+    CopyCpuTensor(input_tensor, output_tensor);
+
+  } else if (input_ort_value->IsTensorSequence()) {
+    const auto* input_tensor_sequence = &input_ort_value->Get<TensorSeq>();
+    auto* output_tensor_sequence = ctx->Output<TensorSeq>(0);
+
+    AllocatorPtr alloc;
+    ORT_RETURN_IF_ERROR(ctx->GetTempSpaceAllocator(&alloc));
+
+    // If the allocation planner had deemed that we re-use the input OrtValue
+    // as the output OrtValue, the pointers of the source TensorSeq and the
+    // target TensorSeq will be the same and the copy is a no-op.
+    // CopySequenceTensor() already has such copy optimizations
+    CopySequenceTensor(alloc, input_tensor_sequence, output_tensor_sequence);
+
+  } else {
+    // Will not reach here
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "Only Optional type OrtValues containing Tensors and Sequence Tensors "
+                           "can be used as input to OptionalGetElement op");
+  }
```

Comment:
Can we use a helper for this logic as it looks the same as what's in Optional::Compute when input_ort_value != nullptr?

Reply:
Good point, thanks